### PR TITLE
Make metadata table for bandchain caching layer

### LINKED ISSUE #829: Make metadata table for bandchain caching layer
The table is a simple KV store to maintain information including ChainID and Latest processed block.

- If the table already exists, the code must check if chain id is correct and panic if not.
- BeginBlocker should begin a transaction.
- DeliverTx should ignore if the height is not latest processed block + 1.
- EndBlocker should update latest processed block and commit.

### PR DESCRIPTION
**Pull request by taobun**
fixed #829

### CODE REVIEW

File: chain/db/table.go [hunk 1]
```diff
@@ -4,6 +4,15 @@ import (
 	"github.com/jinzhu/gorm"
 )
 
+type Metadata struct {
+	Key   string `gorm:"primary_key"`
+	Value string
+}
+
+func (Metadata) TableName() string {
+	return "metatdata"
```
Comment (sorawit) [inline chain/db/table.go]: typo

File: chain/db/table.go [hunk 2]
```diff
@@ -4,6 +4,15 @@ import (
...
+}
+
 type Event struct {
```
Comment (sorawit) [inline chain/db/table.go:12]: Why does `MetaData` have table name, but `Event` does not?

File: chain/db/db.go [hunk 3]
```diff
@@ -10,18 +12,74 @@ import (
...
+func (b *BandDB) ValidateChainID(chainID string) error {
+	var chainIDRow Metadata
+	b.tx.Where("key = ?", "chain-id").First(&chainIDRow)
+	if chainIDRow.Value != chainID {
+		return errors.New("Chain id not match")
+	}
+	return nil
+}
+
+func (b *BandDB) OpenTransaction() {
```
Comment (sorawit) [inline chain/db/db.go]: BeginTransaction

File: chain/app/db_app.go [hunk 4]
```diff
@@ -39,6 +50,18 @@ func (app *dbBandApp) DeliverTx(req abci.RequestDeliverTx) (res abci.ResponseDel
...
+func (app *dbBandApp) BeginBlock(req abci.RequestBeginBlock) abci.ResponseBeginBlock {
+	// Open transaction
+	// if err := app.dbBand.ValidateChainID(app.Ba)
+	app.dbBand.OpenTransaction()
+
 	return app.bandApp.BeginBlock(req)
 }
+
+func (app *dbBandApp) EndBlock(req abci.RequestEndBlock) abci.ResponseEndBlock {
+	res := app.bandApp.EndBlock(req)
```
Comment (sorawit) [inline chain/app/db_app.go]: We need to call `app.bandApp.EndBlock` after we are done with logic. That's where the context is reset.

File: chain/db/db.go [hunk 5]
```diff
@@ -10,18 +12,74 @@ import (
...
+	if b.tx != nil {
+		panic("There is an transaction that left open")
```
Comment (sorawit) [inline chain/db/db.go]: `panic("BeginTransaction: Cannot begin a new transaction without closing the pending one.")`

File: chain/db/db.go [hunk 6]
```diff
@@ -10,18 +12,74 @@ import (
...
+
+func (b *BandDB) RollBack() {
+	b.tx.Rollback()
+	b.tx = nil
+}
+
+func (b *BandDB) GetBlockHeight() int64 {
+	var heightRow Metadata
+	b.tx.Where(
+		Metadata{Key: "height"},
```
Comment (sorawit) [inline chain/db/db.go]: Let's call this "lastProcessedHeight" and define this string as const somewhere.

**Review by sorawit — CHANGES_REQUESTED** [review]

File: chain/app/db_app.go [hunk 7]
```diff
@@ -39,6 +50,18 @@ func (app *dbBandApp) DeliverTx(req abci.RequestDeliverTx) (res abci.ResponseDel
 	return response
 }
 
-func (app *dbBandApp) BeginBlock(req abci.RequestBeginBlock) (res abci.ResponseBeginBlock) {
+func (app *dbBandApp) BeginBlock(req abci.RequestBeginBlock) abci.ResponseBeginBlock {
+	// Open transaction
+	// if err := app.dbBand.ValidateChainID(app.Ba)
+	app.dbBand.OpenTransaction()
+
 	return app.bandApp.BeginBlock(req)
```
Comment (sorawit) [inline chain/app/db_app.go]: Call this first.

**Review by sorawit — CHANGES_REQUESTED** [review]

File: chain/app/db_app.go [hunk 8]
```diff
@@ -39,6 +50,21 @@ func (app *dbBandApp) DeliverTx(req abci.RequestDeliverTx) (res abci.ResponseDel
...
+	}
+
+	return res
+}
+
+func (app *dbBandApp) EndBlock(req abci.RequestEndBlock) abci.ResponseEndBlock {
+	res := app.bandApp.EndBlock(req)
+	// Do other logic
+	app.dbBand.SetBlockHeight(req.GetHeight())
+	app.dbBand.Commit()
```
Comment (sorawit) [inline chain/app/db_app.go:88]: This should be during `Commit`, not during `EndBlock`.

**Review by sorawit — CHANGES_REQUESTED** [review]

File: chain/app/db_app.go [hunk 9]
```diff
@@ -25,20 +25,52 @@ func NewDBBandApp(
 	return &dbBandApp{bandApp: app, dbBand: dbBand}
 }
 
+func (app *dbBandApp) InitChain(req abci.RequestInitChain) abci.ResponseInitChain {
+	app.dbBand.BeginTransaction()
+
+	app.dbBand.SaveChainID(req.GetChainId())
+	app.dbBand.SetBlockHeight(0)
```
Comment (sorawit) [inline chain/app/db_app.go]: SetLastProcessedHeight

Comment (sorawit) [reply chain/app/db_app.go]: Also Get...

**Review by sorawit — CHANGES_REQUESTED** [review]

**Review by sorawit — APPROVED** [review]

**Review by evilpeach — APPROVED** [review]

**Pull request merged**